h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,174 @@
+package ai.h2o.targetencoding;
+
+import water.MRTask;
+import water.fvec.CategoricalWrappedVec;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+
+class BroadcastJoinForTargetEncoder {
+
+  static class FrameWithEncodingDataToArray extends MRTask<FrameWithEncodingDataToArray> {
+    
+    /**
+     * Numerators and denominators are being stored in the same row-array of the 2D array.
+     * 
+     *        _encodingDataPerNode[k][0 .. numOfCatLevels -1] will be storing numerators
+     *        _encodingDataPerNode[k][numOfCatLevels .. (2*numOfCatLevels -1)] will be storing denominators,
+     *        
+     *        where k is a current fold value (folds = {0,1,2,3...k}) or 0 when _foldColumnIdx == -1; 
+     */
+    int[][] _encodingDataPerNode = null;
+    
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx, _cardinalityOfCatCol, _maxFoldValue;
+
+    FrameWithEncodingDataToArray(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx, int cardinalityOfCatCol, int maxFoldValue) {
+      _categoricalColumnIdx = categoricalColumnIdx;
+      _foldColumnIdx = foldColumnId;
+      _numeratorIdx = numeratorIdx;
+      _denominatorIdx = denominatorIdx;
+      _cardinalityOfCatCol = cardinalityOfCatCol;
+
+      if(foldColumnId == -1) {
+        _encodingDataPerNode = new int[1][_cardinalityOfCatCol * 2];
+      } else {
+        assert maxFoldValue >= 1 : "There should be at leas two folds in the fold column";
+        assert _cardinalityOfCatCol > 0 && _cardinalityOfCatCol < (Integer.MAX_VALUE / 2)  : "Cardinality of categ. column should be within range (0, Integer.MAX_VALUE / 2 )";
+        _encodingDataPerNode = new int[maxFoldValue + 1][_cardinalityOfCatCol * 2];
+      }
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        int levelValue = (int) categoricalChunk.at8(i);
+
+          // We are allowed to do casting to `int` as we have validation before submitting this MRTask
+          int foldValue = _foldColumnIdx != -1 ? (int) cs[_foldColumnIdx].at8(i) : 0;
+          int[] arrForNumeratorsAndDenominators = _encodingDataPerNode[foldValue];
+
+        synchronized (arrForNumeratorsAndDenominators) { // Sync on whole array as there is no way to sync on two objects ( two rows of 2D array)
```

Comment:
@deil87 trying to figure out why this would be necessary: modifying the same value looks a bit alien to the map-reduce approach. Here you don't care if 2 parallel maps are modifying the `_encodingDataPerNode` 2d array concurrently as soon as they're not modifying the same cell.
But I don't see how it should be possible here as you never read the previous content of the cell. I'm missing something?

Reply:
@seb-h2o first thing... I forgot to remove comment and it is outdated. 

> modifying the same ref looks a bit alien to the map-reduce approach

In a multinode scenario refs will be different. 

> But I don't see how it should be possible here as you never read the previous content of the cell. I'm missing something?

Here we are reading row/array
```
int[] arrForNumeratorsAndDenominators = _encodingDataPerNode[foldValue];
```
Here we are modifying
```
          arrForNumeratorsAndDenominators[levelValue] = (int) numeratorChunk.at8(i);
          arrForNumeratorsAndDenominators[_cardinalityOfCatCol + levelValue] = (int) denominatorChunk.at8(i);
```
So I think there is an opportunity for race condition for the maps that are being run concurrently on the same mode.  
Another question is whether we operate on the same data( cells) as you have said. And it looks like we are operating with just the cells of the array for a given  fold and we should be fine even without sync. But I have test that would fail without this sync block(if you run this test multiple times) - `TargetEncoderBuilderTest.transform_KFold_scenario`. 

I think that the reason for race condition is that reduce phase might be intertwined with map phase and as we read and write in reduce method we need sync there as well. Do we wait for all maps to finish and only then run reduce phase? From my understanding( debugging) we are not waiting.

Reply:
@deil87 ref was a bad wording sorry, I meant instance, and anyway `synchronized` is no use in multinode.
Yes, I noticed the global read/write, but what for me would have justified the `synchronized` is sth like:
```
arrForNumeratorsAndDenominators[levelValue] += (int) numeratorChunk.at8(i);
```

> Do we wait for all maps to finish and only then run reduce phase? From my understanding( debugging) we are not waiting.

You're right, F/J applies reduce for every fork, no wait.
The sync is necessary here not because of the `map` (we both agree that only one thread can change a given cell), but because reduce is modifying this 2d array in-place.
If your `FrameWithEncodingDataToArray` has 2 fields, one for map, and one for the result of `reduce`, you wouldn't need the `sync`: I'd be curious about the performance impact

Reply:
@seb-h2o 
> ref was a bad wording sorry, I meant instance, and anyway

I think that both refs and instances will be different. We fill this array(or better these arrays) on different nodes independently.

> and anyway synchronized is no use in multinode.

I know that `synchronized` is not working across nodes because it can't work through serialisation/deserialisation processes. But here `synchronized` is needed(it is not actually needed if we use another approach - see below) when we operate within one node and, as I mentioned before, our `map` and `reduce` methods are getting intertwined. So reduce in multinode is  being done both locally and between instances from different nodes 

> If your FrameWithEncodingDataToArray had 2 encodingData fields, one for map, and one for the result of reduce, you wouldn't need the sync: I'd be curious about the performance impact

We would need to have twice as much memory and for cases when we have really high cardinalities and also many fold values this will be quite memory expensive. 

Gotcha! I think I come up with a better solution. Instead of all synchronisations or extra arrays 
we can do following in reduce method:
```
if (leftArr != rightArr) {
            int valueFromLeftArr = leftArr[rowIdx][colIdx];
            int valueFromRIghtArr = rightArr[rowIdx][colIdx];
            leftArr[rowIdx][colIdx] = Math.max(valueFromLeftArr, valueFromRIghtArr);
}
```
as when we are reducing lets call it within node we don't need to merge arrays at all, and we will not have no concurrency issues as we don't do anything

Reply:
@deil87 I think we have a misunderstanding with the ref/instance thing: of course I know they should be different, but apparently they weren't always different, otherwise you wouldn't have needed the `synchronized` lock.
Now can you please explain your fix?
If you add a condition `if (leftArr != rightArr)`, it's because you suggest that they could be equal... which is exactly what I was pointing at.
M/R should try to avoid that imo.

Regarding the memory usage, yes, my suggestion would temporarily double the memory usage for the 2D array, but once the reduction is done, you can reassign it to the result array.
There's always a trade-off between memory and perf, in this case, I would happily favour perf against memory. You have a 2D int array, |cat| * 2 * |folds| * 4 bytes < 1 MB for 10 folds, 10000 categories (before, you were creating 3*100000 locks for the same map/reduce operation).

Actually, I looked into details at the `MRTask` impl, and I didn't expect it to (shallow) clone the original instance for both left and right map operations, I guess that's why you can have `leftArr == rightArr` in the reduce operation (I find this to be very against M/R model, which is supposed to avoid those kind of conflicts).
And then, your latest change makes sense somehow: as the result is shared by lhs and rhs within node, then the reduction is actually done during mapping... not very M/Rish but efficient.

Btw, if you know in advance that `if (leftArr != rightArr)`, you can as well move the condition before the loop, and completely skip the reduce within node, it would be more clear.
Also please add comment explaining why: this really works because of the implementation details of `MRTask`. Usually, when reducing to `Frame`, this is handled by the `MRTask.map(Chunk, NewChunk)` methods. I believe that reducing to arrays is unusual in h2o-3, that's why those races arose.

Reply:
Also, as you said that categorical cardinality can be high (for folds, I don't think users have enough time in their life to use TE with 100 xval), maybe you should use `MemoryManager.malloc` then when creating your arrays.

Reply:
@seb-h2o 
> I think we have a misunderstanding with the ref/instance thing

Yes, there was a misunderstanding. I'm sorry. Now I see what you were referring to.

> M/R should try to avoid that imo

Agree, maps should be pure and don't share same instance. In given use case we don't have conflicts during map phase as we work with different cells and there might not be any collisions. But due to an ability to reduce before all maps are done we can have race conditions. Current design of MRTask has its benefits of saving memory during computations....but we pay for it with higher carefulness during coding. It is always the case when you share something :)

> There's always a trade-off between memory and perf, in this case, I would happily favour perf against memory

I would not favour any of them. As it is called "trade-off" it by definition always depends on the situation. It is not number of folds that might be high, most likely it is number of categorical levels. I had mentioned two terms because even having relatively small number of folds it affects total value significantly due to a multiplication operation between them. Toy example: 0.5Gb for cat. levels and having 10 folds will make it 5Gb.

> Actually, I looked into details at the MRTask impl, and I didn't expect it to (shallow) clone the original instance for both left and right map operations, I guess that's why you can have `leftArr == rightArr` in the reduce operation (I find this to be very against M/R model, which is supposed to avoid those kind of conflicts).

Yes, I also had a look into implementation details of our MR framework. With shallow copy we can't do check `leftArr != rightArr` before  `reduce` is being called  as MRTask instances are always different but internal fields referencing the same instances(in case of MRTasks from the same node) and we don't know particular fields of MRTask of course => we can only check inside reduce method. 

>  then the reduction is actually done during mapping... not very M/Rish but efficient.

 Yes, feels like this. 

> Btw, if you know in advance that if (leftArr != rightArr), you can as well move the condition before the loop, and completely skip the reduce within node, it would be more clear.

Yes, good catch! 

> I believe that reducing to arrays is unusual in h2o-3, that's why those races arose.

I think that it is not about arrays. It is about mutable vs immutable approach. If you do it in immutable way with new NewChunk `MRTask.map(Chunk, NewChunk)` then no race conditions. If you do `MRTask.map(Chunk[])` then RCs are very welcome.
I have seen reducing to arrays in AstGroup functionality. Actually if you take a look at the comment for `GBTaskAggsPerNode.reduce` method you will see quite similar story. 

> MemoryManager.malloc then when creating your arrays

I was considering to use `MemoryManager` but there were no method for 2D int array. But once you are also suggesting this I've decided to add one. Btw, there is a code duplication between `MemoryManager.malloc8d` and `CoxPHUtils.malloc2DArray`. Also `CoxPHUtils.malloc3DArray` maybe could be moved to `MemoryManager` so we can get rid of `CoxPHUtils`.

Thank you for the discussion! I think It was a good opportunity for both of us to learn more about our MRTask framework.

Reply:
@seb-h2o added comment for check in reduce method. Let me know if it is not clear. Also created separate PR for malloc4 https://github.com/h2oai/h2o-3/pull/3994/files

Reply:
@deil87 definitely useful conversation, thanks.

Maybe for future improvement, I'm still wondering if there's a way to avoiding passing the full 2D array to the forked tasks. It probably requires support at `MRTask` level in a similar way to what is done with `MRTask.map(Chunk, NewChunk)` though...

> I think that it is not about arrays. It is about mutable vs immutable approach

Yeah, totally agree, I just meant that if reducing to arrays was more common, someone would have implemented at `MRTask` level to avoid those races, like mentioned above.